protocol/0079-TGAS-transaction_gas_costs.md

```diff
@@ -0,0 +1,112 @@
+- `network.transaction.defaultgas` is a network parameter with type unsigned integer with a minimum value of `1` and maximum value of `99` and default value of `1`. If the parameter is changed through governance then the next block after enactment will respect the new `defaultgas`. 
+
+Note that the min / max values set above are deliberate: as we'll see below we can fit at least one transaction with default gas into a block. 
+
+## Including transactions 
+
+Each transaction will have a gas cost assigned to it. Any transaction not specifically named below has gas cost of `network.transaction.defaultgas`. 
+The consensus layer (Tendermint) will choose transactions from the proposer's mempool to include into a block with a maximum total cost of `network.transactions.maxgasperblock`. 
+It must *not* try to include the transactions with the highest gas; in fact it should include transactions in the sequence it's seen them until the block gas limit is reached. 
+
+## Default transactions cost
+
+Every transaction not listed below will have gas cost `network.transaction.defaultgas`. 
+
+
+## Dynamic transactions costs
+
+Cost of transaction depends mainly on the state of underlying market and below we set out costs of transactions based on market state. 
+Vega will capture the needed statistical variables (see below) on a per-market basis (or per-whatever if other dynamically costed transactions are added, for now per-market is sufficient) from the previous block so that they don't need to be looked up dynamically during block creation. 
+
+Variables needed:
+- `network.transactions.maxgasperblock` - `maxGas`
+- number of price levels on the order book taken, this can count just static volume or static plus dynamic(*) - `levels`
+- number of pegged orders - `pegs`
+- number of LP shape levels on the market - `shapes` (this means that if LP A has 3 pegs on buy and 2 on sell and LP B has 1 buy peg and 4 sell pegs and they're the only LPs then this value is 10 ) 
+- number of positions on the market - `positions`
+
+(*) update after implementation
+
+Constants needed:
+- `peg cost factor = 50` nonnegative decimal
+- `LP shape cost factor = 100` nonnegative decimal
+- `position factor = 1` nonnegative integer
+- `level factor = 0.1` nonnegative integer 
+- `batchFactor = 0.5` decimal between `0.1 and 0.9`.
+
+### Any type of limit or market order
+
+```
+gasOrder = network.transaction.defaultgas + peg cost factor x pegs 
+                                        + LP shape cost factor x shapes 
+                                        + position factor x positions 
+                                        + level factor x levels
+gasOrder = min(maxGas-1,gasOrder)
+```
+
+### Cancellation of any order
+
+```
+gasCancel = network.transaction.defaultgas + peg cost factor x pegs 
```

Comment:
We need a cost for the two special mass cancel tx types:

- cancel all orders on a market for a party = same cost as cancelling ONE order on that market (we only do any other processing like reprice once ALL are removed)
- cancel all orders on all markets for a party = same cost as cancelling ONE order on each market the party has orders on

Reply:
✅